src/relax/backend/vm/rewrite_cuda_graph.cc

```diff
@@ -0,0 +1,408 @@
+    Array<StructInfo> alloc_tensors_struct_info;
+    for (const auto& var : scope_.graph.alloc_tensors) {
+      region->alloc_tensor_to_index.emplace_back(var, alloc_tensors.size());
+      alloc_tensors.push_back(builder->Emit(scope_.graph.bindings.at(var)->value));
+      alloc_tensors_struct_info.push_back(
+          Downcast<StructInfo>(GetRef<Var>(var)->struct_info_.value()));
+    }
+    auto output = builder->Emit(Tuple(alloc_tensors));
+    auto block = builder->EndBlock();
+    auto func_body = builder->Normalize(SeqExpr({block}, output));
+    auto func = Function({}, func_body, TupleStructInfo(alloc_tensors_struct_info));
+    return func;
+  }
+
+  Function EmitStaticGraph(BlockBuilderNode* builder, StaticRegion* region) {
+    Var allocs{"allocs", region->alloc_func->ret_struct_info};
+    builder->BeginBindingBlock();
+    for (int i = 0; i < static_cast<int>(scope_.graph.alloc_tensors.size()); ++i) {
+      const VarNode* tensor = scope_.graph.alloc_tensors[i];
+      const Expr alloc = builder->Normalize(TupleGetItem(allocs, i));
+      builder->EmitNormalized(VarBinding(GetRef<Var>(tensor), alloc));
+    }
+    for (const VarNode* node : scope_.graph.nodes) {
+      const VarBindingNode* binding = scope_.graph.bindings.at(node);
+      builder->EmitNormalized(GetRef<VarBinding>(binding));
+    }
+    auto void_output = builder->Emit(Tuple(Array<Expr>{}));
+    auto block = builder->EndBlock();
+    auto func_body = builder->Normalize(SeqExpr({block}, void_output));
+    auto func = Function({allocs}, func_body, TupleStructInfo(Array<StructInfo>{}));
+    return func;
+  }
+
+  void SummarizeStaticRegion() {
+    PruneOutputNodes(&scope_.graph, outputs_);
+
+    // Find all bindings other than alloc_tensors to emit
+    std::vector<VarBinding> bindings_to_emit;
+    bool has_kernel_launch = false;
+    for (const auto* var : scope_.graph.nodes) {
+      auto it = scope_.graph.bindings.find(var);
+      const VarBindingNode* binding = it->second;
+      if (const CallNode* call = binding->value.as<CallNode>();
+          call != nullptr && call->op.as<GlobalVarNode>()) {
+        has_kernel_launch = true;
+      }
+      bindings_to_emit.push_back(GetRef<VarBinding>(binding));
+    }
+    if (!has_kernel_launch) return;
+
+    BlockBuilder static_region_builder = BlockBuilder::Create(mod_);
+
+    StaticRegion region;
+    // Emit alloc_tensors as a separate function
+    region.alloc_func = EmitStaticAllocations(static_region_builder.operator->(), &region);
+    // Emit the rest of the graph, with original alloc_tensors replaced with the function parameters
+    region.capture_func = EmitStaticGraph(static_region_builder.operator->(), &region);
+    region.launch_point = scope_.graph.bindings.at(scope_.graph.nodes.front());
+    region.bindings = std::move(scope_.graph.bindings);
+
+    scope_.static_regions.push_back(std::move(region));
+    scope_.graph = DependencyGraph();
+  }
+
+  IRModule mod_;
```

Comment:
Please keep the member vars private. Additionally, dispatched `VisitExpr_`s can be private too